snow/engine/snowman/transitive.go

```diff
@@ -473,41 +478,95 @@ func (t *Transitive) Start(ctx context.Context, startReqID uint32) error {
+	} else {
+		preferredID, err := t.getInitialPreferenceVM.GetInitialPreference(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to get initial preference: %w", err)
+		}
+
+		preferredBlock, err = t.VM.GetBlock(ctx, preferredID)
+		if err != nil {
+			return fmt.Errorf("failed to get initial preferred block: %w", err)
+		}
+	}
+
+	// Check if the initial setup is more recent than the last accepted
+	// block. This can occur if we shut down while a set of blocks was still
+	// processing in consensus.
+	preferredTip := lastAcceptedID
+	if preferredBlock.Height() > lastAccepted.Height() {
+		preferredTip = preferredID
+	}
+
+	preferredChain := make([]snowman.Block, 0)
+	for preferredBlock.Status() == choices.Processing {
+		// During bootstrap our accepted tip can grow, so we should only
+		// re-issue blocks into consensus with a greater height than what
+		// we currently have.
+		if preferredBlock.Height() > lastAccepted.Height() {
+			preferredChain = append(preferredChain, preferredBlock)
+		}
+
+		preferredBlock, err = t.VM.GetBlock(ctx, preferredBlock.Parent())
+		if err != nil {
+			return err
+		}
+	}
+
+	// reverse the preferred chain so the blocks are now in chronological order
+	slices.Reverse(preferredChain)
+
 	// initialize consensus to the last accepted blockID
 	lastAcceptedHeight := lastAccepted.Height()
 	if err := t.Consensus.Initialize(t.Ctx, t.Params, lastAcceptedID, lastAcceptedHeight, lastAccepted.Timestamp()); err != nil {
 		return err
 	}
 
+	issuedMetric := t.metrics.issued.WithLabelValues(builtSource)
 	// to maintain the invariant that oracle blocks are issued in the correct
 	// preferences, we need to handle the case that we are bootstrapping into an oracle block
 	if oracleBlk, ok := lastAccepted.(snowman.OracleBlock); ok {
 		options, err := oracleBlk.Options(ctx)
 		switch {
 		case err == snowman.ErrNotOracle:
 			// if there aren't blocks we need to deliver on startup, we need to set
-			// the preference to the last accepted block
-			if err := t.VM.SetPreference(ctx, lastAcceptedID); err != nil {
-				return err
-			}
+			// the preference to the preferred tip
 		case err != nil:
 			return err
 		default:
-			issuedMetric := t.metrics.issued.WithLabelValues(builtSource)
 			for _, blk := range options {
 				// note that deliver will set the VM's preference
 				if err := t.deliver(ctx, t.Ctx.NodeID, blk, false, issuedMetric); err != nil {
 					return err
 				}
 			}
 		}
-	} else if err := t.VM.SetPreference(ctx, lastAcceptedID); err != nil {
+	}
+
+	// Re-issue all blocks in the preferred chain into consensus
+	for _, preferredBlk := range preferredChain {
+		if err := t.deliver(ctx, t.Ctx.NodeID, preferredBlk, false, issuedMetric); err != nil {
+			return err
+		}
+	}
+
+	if err := t.VM.SetPreference(ctx, preferredTip); err != nil {
 		return err
 	}
 
 	t.Ctx.Log.Info("starting consensus",
-		zap.Stringer("lastAcceptedID", lastAcceptedID),
-		zap.Uint64("lastAcceptedHeight", lastAcceptedHeight),
```

Comment:
I think we still want to be logging the heights here (They are generally much more readable than the IDs).